pyro/distributions/empirical.py

```diff
@@ -0,0 +1,160 @@
+
+    def add(self, value, log_weight=None, weight=None):
+        """
+        Adds a new data point to the sample. The values in successive calls to
+        ``add`` must have the same tensor shape and size. Optionally, an
+        importance weight can be specified via ``log_weight`` or ``weight``
+        (default value of `1` is used if not specified).
+
+        :param torch.Tensor value: tensor to add to the sample.
+        :param torch.Tensor weight: log weight (optional) corresponding
+            to the sample.
+        :param torch.Tensor log_weight: weight (optional) corresponding
+            to the sample.
+        """
+        if is_validation_enabled():
+            if weight is not None and log_weight is not None:
+                raise ValueError("Only one of ```weight`` or ``log_weight`` should be specified.")
+            if torch.is_tensor(weight) and weight.dim() > 0:
+                raise ValueError("``weight.dim() > 0``, but weight should be a scalar.")
+
+        weight_type = torch.Tensor if value.dtype in (torch.int32, torch.int64) else value.type()
+        # Apply default weight of 1.0.
+        if log_weight is None and weight is None:
+            log_weight = torch.tensor(1.0).type(weight_type).log()
+        elif weight is not None and log_weight is None:
+            log_weight = math.log(weight)
+        if isinstance(log_weight, numbers.Number):
+            log_weight = torch.tensor(log_weight).type(weight_type)
+
+        # Seed the container tensors with the correct tensor types
+        if self._samples is None:
+            self._samples = value.new()
+            self._log_weights = log_weight.new()
+        # Append to the buffer list
+        self._samples_buffer.append(value)
+        self._weights_buffer.append(log_weight)
+
+    def sample(self, sample_shape=torch.Size()):
+        self._finalize()
+        idxs = self._categorical.sample(sample_shape=sample_shape)
+        return self._samples[idxs]
+
+    def log_prob(self, value):
+        """
+        Returns the log of the probability mass function evaluated at ``value``.
+        Note that this currently only supports scoring single values and not
+        a tensor of values.
+
+        :param torch.Tensor value: scalar or tensor value to be scored.
+        """
+        if is_validation_enabled():
+            if value.size() != self.event_shape:
+                raise ValueError("``value.size()`` must be {}".format(self.event_shape))
+        self._finalize()
+        selection_mask = self._samples.eq(value).contiguous().view(self.sample_size, -1)
+        # Return -Inf if value is outside the support.
+        if not selection_mask.any():
+            return self._log_weights.new_zeros(torch.Size()).log()
+        idxs = torch.arange(self.sample_size)[selection_mask.min(dim=-1)[0]]
+        log_probs = self._categorical.log_prob(idxs)
+        return log_sum_exp(log_probs)
+
+    @property
+    def batch_shape(self):
```

Comment:
I think you can omit this property this if you pass the `batch_shape` arg to `super(..., self).__init__()`

Reply:
Will change.